### PR TITLE
[Docker] Move to using DGL 2.2 for local image, update CUDA source image to supported version, add fetch_and_run.sh script for local container

### PR DESCRIPTION
**Pull request by thvasilo**
*Issue #, if available:*

*Description of changes:*

* We move version of DGL to 2.2.1 for the local image
* We add fetch_and_run.sh as the command for the local image, to allow easy execution of arbitrary scripts through AWS Batch etc. See https://aws.amazon.com/blogs/compute/creating-a-simple-fetch-and-run-aws-batch-job/ for details
* Update CUDA image to 12.1.1, 12.1.0 is EOL.
* Move local Dockerfile to its own directory to keep it consistent with the other Dockerfiles (sagemaker, parmetis)
* Add script to easily push GSF images to ECR.

## Testing

Ran multiple NC jobs with ogbn-mag, ACM example data, ogb-paper-100M on up to 4 GPU workers.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/local/fetch_and_run.sh [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script can help you download and run a script from S3 using aws-cli.
+# It can also download a zip file from S3 and run a script from inside.
+# See below for usage instructions.
```
Comment (jalencato) [inline docker/local/fetch_and_run.sh:17]: May we add https://aws.amazon.com/blogs/compute/creating-a-simple-fetch-and-run-aws-batch-job/ to this PR description for how to use the script here?

**Review by jalencato — COMMENTED** [review]
Some minor comment and questions

File: docker/local/fetch_and_run.sh (hunk 1 above)
Comment (jalencato) [reply docker/local/fetch_and_run.sh:17]: Sorry, I see it is already in the description. I will mark it as resolved.

File: docker/build_docker_oss4local.sh [hunk 2]
```diff
@@ -33,11 +35,16 @@ fi
...
+cp $SCRIPT_DIR"/local/fetch_and_run.sh" $GSF_HOME"/docker/code/"
 cp -r $GSF_HOME"/python" $GSF_HOME"/docker/code/python"
 cp -r $GSF_HOME"/examples" $GSF_HOME"/docker/code/examples"
 cp -r $GSF_HOME"/inference_scripts" $GSF_HOME"/docker/code/inference_scripts"
 cp -r $GSF_HOME"/tools" $GSF_HOME"/docker/code/tools"
 cp -r $GSF_HOME"/training_scripts" $GSF_HOME"/docker/code/training_scripts"
+# If using a local DGL installation we expect it to be one level above GSF
+# mkdir -p code/dgl
+# rsync -qr "${GSF_HOME}/../dgl/" code/dgl/ --exclude .venv --exclude dist --exclude ".*/" \
+#         --exclude "*__pycache__" --exclude "third_party"
```
[the remark this answers is not in the document]
Comment (classicsong) [reply docker/build_docker_oss4local.sh:47]: Agree with jalencato

File: docker/build_docker_oss4local.sh [hunk 3]
```diff
@@ -62,7 +69,7 @@ fi
 DOCKER_BUILDKIT=1 docker build \
     --build-arg DEVICE=$DEVICE_TYPE \
     --build-arg SOURCE=${SOURCE_IMAGE} \
-    -f "${GSF_HOME}/docker/Dockerfile.local" . -t $DOCKER_FULLNAME
+    -f "${GSF_HOME}/docker/local/Dockerfile.local" . -t $DOCKER_FULLNAME
```
Comment (classicsong) [inline docker/build_docker_oss4local.sh:72]: What is the difference between docker/local/Dockerfile.local and docker/Dockerfile.local?

File: docker/local/fetch_and_run.sh [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script can help you download and run a script from S3 using aws-cli.
```
Comment (classicsong) [inline docker/local/fetch_and_run.sh:15]: What is the purpose of this file?

File: docker/build_docker_oss4local.sh (hunk 3 above)
Comment (thvasilo) [reply docker/build_docker_oss4local.sh:72]: Forgot to delete the old version, I'm just moving the local Dockefile to its own directory, so it matches the layout of sagemaker and parmetis that also exist in their own dir.

File: docker/local/fetch_and_run.sh (hunk 4 above)
Comment (thvasilo) [reply docker/local/fetch_and_run.sh:15]: Mentioned in the PR descirption

>We add fetch_and_run.sh as the command for the local image, to allow easy execution of arbitrary scripts through AWS Batch etc. See https://aws.amazon.com/blogs/compute/creating-a-simple-fetch-and-run-aws-batch-job/ for details

For someone not using AWS Batch this should not make a difference in how they use the local containers, but we can use this in Batch to pull scripts from S3 and run them. Currently any script needs to pre-exist on the image, which means that we need to pull a new image every time we modify e.g. a training script.

File: docker/build_docker_oss4local.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (thvasilo) [reply docker/build_docker_oss4local.sh:47]: I'll move it to the README here as we only expect devs to use this.

**Comment by thvasilo** [comment]
Applied review comments and clarified questions, let me know if there's anything else needed @classicsong @jalencato

File: docker/README.md [hunk 5]
```diff
@@ -27,22 +27,54 @@ Once you have the GraphStorm repository cloned, please use the following command
...
+```bash
+bash docker/push_gsf_container.sh graphstorm local-gpu us-west-2 1234567890
+```
+
+## Using a customer DGL codebase
+---------------
+To use a local DGL codebase, you'll need to modify the build script and Dockerfile.local.
+
+
+You can add the followingto the build_docker_oss4local.sh:
```
Comment (jalencato) [inline docker/README.md:65]:
```suggestion
You can add the following to the build_docker_oss4local.sh:
```

File: docker/README.md [hunk 6]
```diff
@@ -27,22 +27,54 @@ Once you have the GraphStorm repository cloned, please use the following command
...
-There are three arguments of the `build_docker_oss4local.sh`:
+There are four arguments of the `build_docker_oss4local.sh`:
 
 1. **path-to-graphstorm**(required), is the absolute path of the "graphstorm" folder, where you
 cloned the GraphStorm source code. For example, the path could be "/code/graphstorm".
 2. **docker-name**(optional), is the assigned name of the to be built Docker image. Default is
 "graphstorm".
 3. **docker-tag**(optional), is the assigned tag name of the to be built docker image. Default is
 "local".
+3. **device**(optional), is the intended execution device for the image. Can be `cpu` or `gpu`, default is
```
Comment (jalencato) [inline docker/README.md:41]:
```suggestion
4. **device**(optional), is the intended execution device for the image. Should be one of `cpu` or `gpu`, default is
```

**Review by jalencato — APPROVED** [review]
LGTM, some small typos

**Pull request merged**